### ISSUE TITLE
Does it make sense to create the reference inside the driver?

### ISSUE THREAD

**Issue by ntilwalli** [dralletje/cycle-firebase#1]
Thanks for making this driver!  I just started taking a look and I have an initial question.

I noticed that your README example expects users to import the firebase library directly and call `new Firebase(...)` directly while constructing the sink.  Do you think it makes sense to include it as a peer dependency of your driver and manage the ref creating internally?  You could simply set the database name as the the argument instead of the direct firebase instance reference, right?  That way internal details could be hidden.

I could def make a pull request, I just wanted to get your thoughts.

**Comment by dralletje** [comment]
I actually had that in the beginning, so that `makeFirebaseDriver` would accept the firebase url instead of the firebase ref, but that was harder to test (pass in custom firebase ref). However, I think we could just allow both! :smile: So if a string is passed in, create a baseref, and if not... use the things you get as baseref (maybe some sanity checks to make sure it has the required functions?).

And I'd love a pull request! That is so cool o_o

Firebase should be a normal dependency I think though, as when you are passing in just the string, you shouldn't be bothered with installing the dependency as well, I'd say :) (But I'd love your thoughts on that :D)

**Comment by ntilwalli** [comment]
Cool.  No I agree, it's a normal dependency, dunno why I wrote peer. :)  Another question.  What are your thoughts on authentication?  As it currently stands I think I'd have to access the base ref directly and call the authentication functions directly on it in my application code which would qualify as side-effects or am I misunderstanding something?  Is there another way?

**Comment by dralletje** [comment]
Ohhh yes, I only thought about _getting_ the authentication info, dang! Honestly, I have no idea where to put that :joy:

**Comment by dralletje** [comment]
Oh dang, I don't have a way to push to a list now as well, nooo!!! :D

**Comment by ntilwalli** [comment]
Ha. :) Maybe messages could be passed to the driver which indicate the behavior the user wants in a `type` attribute which could be used for filtering in the driver and have a `payload`/`data` attribute that could then be used based on the type?  LIke `{type: "set", data: "blah"}`, `{type: "authWithOAuthPopup", data: {provider: "facebook", authHandler: <some function>}}` I'm just throwing ideas out there. Should work, but I wonder if there is a more idiomatic way.

**Comment by dralletje** [comment]
What about https://github.com/dralletje/cycle-firebase/commit/78183ab6bce2e90f7b7abcbccd5a7e8bd69a7c7d for pushing? ^^

**Comment by ntilwalli** [comment]
I'm still kind of new to Firebase so I'm not clear on when I would need to manually create a key for pushing.  Doesn't calling `push` automatically generate one of these keys for you?  Are there use cases for manual generation?  If the message approach I proposed above with a message like `{type: "push", data: {...}}` was used, would that bypass the need?

**Comment by dralletje** [comment]
Firebase does generate the pushId's for you if you call `.push`, but you most of the time want to have those Id's after you push it, so you can manipulate it further or show it to the user. If you want to do a same thing with the message channel, you'll have to listen for the result of push again: 

```
// Listen for pushId's to be broadcasted again
let userId$ = firebase.pushedTo(`users`)

return {
  //  Just to show we are pushing
  firebase: Observable.just({
    actions: firebase.$pushTo(`users`, {...})
  })
}
```

but what if we want to push multiple users at the same time, and don't want to treat all the items pushed to `users` the same?

Also, I don't want to give up the declarative approach, where I find stream list of actions not declarative :p

The pushIds are generated exactly like firebases now, by the way :-) https://gist.github.com/mikelehen/3596a30bd69384624c11

**Comment by ntilwalli** [comment]
I see. I think the `pushId$` is valuable and looks good.  Unrelated to the `pushId$`, I wonder if Firebase lends itself to being declarative.  I don't think it does, especially as it relates to data being pushed to the sink.  My client is not the only client that can update my Firebase, which is different than the DOM.  I have full control over the DOM all the time, which means it's possible to pairwise diff for changes since I'm definitely the one that made the changes.  That does not apply to Firebase since any number of clients can update the data and the current state (as stored in `pairwise`) will become invalid the moment a data update happens that wasn't caused by me.  Isn't that correct?  Which would mean the next time a sink write happens which reflects the latest Firebase state, the diffed changes will include redundant items (that are already currently in Firebase) and will try to re-write them.  Is my interpretation wrong?

**Comment by ntilwalli** [comment]
In the above comment I'm assuming components are the only subscribers to the data sources.  If the driver also subscribes to the data changes and updates it's current data state as well, then a diffing approach could work.

**Comment by dralletje** [comment]
No, you interpretation is right. Still thinking of a way to get that working, as I think having a stream of commands/actions is totally uncycle-ish. Yes, it is moving the side effects to the edge of the application, but in no way it is mimicking the 'tree structure declarative way of doing things' that makes react so awesome ;)

I assume a app would listen to the data it cares about setting multiple times, but maybe we can make that obligatory. Or indeed, listening to the data that is currently in the graph, so you know when to update even if the data didn't change on the client.

Lots of challenges, but I'm not going to make concessions here.

Making a command stream based driver is way too easy too ;)

**Comment by ntilwalli** [comment]
@dralletje I've sketched out a different Firebase driver which is modeled a bit on `cycle-dom` in the sense that it uses `select` and `events` semantics and offers a form of hierarchical namespaces. It also mixes in the command approach I discussed but by default commands requests/responses can be isolated via their namespace/path so a requester only sees the responses associated with their own requests.   I'd love your feedback.  It's at https://github.com/ntilwalli/cycle-fire
